File: app/src/main/java/com/constraint/vagabond/main/MainPresenterImpl.java
package com.constraint.vagabond.main;

import com.constraint.vagabond.data.RecreationalArea;

import java.util.ArrayList;
import java.util.List;

public class MainPresenterImpl
    implements MainContract.presenter, MainContract.GetRecAreasInteractor.OnFinishedListener {

  private MainContract.MainView mainView;
  private MainContract.GetRecAreasInteractor getRecAreasInteractor;

  public MainPresenterImpl(
      MainContract.MainView mainView, MainContract.GetRecAreasInteractor getRecAreasInteractor) {

    this.mainView = mainView;
    this.getRecAreasInteractor = getRecAreasInteractor;
  }

  @Override
  public void onDestroy() {
    mainView = null;
  }

  @Override
  public void onSearch(String s) {
    if (mainView != null) {
      mainView.showProgress();
    }
    getRecAreasInteractor.getRecAreasList(this, s);
  }

  @Override
  public void onFinished(List<RecreationalArea> recAreasList) {
    if (mainView != null) {
      mainView.setDataToRecyclerView(recAreasList);
      mainView.hideProgress();
    }
  }

  @Override
  public void onFailure(Throwable t) {
    if (mainView != null) {
      mainView.onResponseFailure(t);
      mainView.hideProgress();
    }
  }
}

File: app/src/main/java/com/constraint/vagabond/main/search/MainPresenterImpl.kt
package com.constraint.vagabond.main.search

import com.constraint.vagabond.data.entities.RecreationalAreaList
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableSingleObserver
import io.reactivex.schedulers.Schedulers

class MainPresenterImpl(
        private var mainView: MainContract.MainView?) : MainContract.presenter {

    private val compositeDisposable = CompositeDisposable()

    override fun onDestroy() {
        compositeDisposable.dispose()
        mainView = null
    }

    override fun onSearch(s: String, apiKey: String) {
        if (mainView != null) {
            mainView!!.showProgress()
        }
        val repository = SearchRepositoryProvider.provideSearchRepository()
        compositeDisposable.add(repository
                .getRecAreasList(s, apiKey)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(object : DisposableSingleObserver<RecreationalAreaList>() {
                    override fun onSuccess(t: RecreationalAreaList) {
                        mainView?.setDataToRecyclerView(t)
                        mainView?.hideProgress()
                    }

                    override fun onError(e: Throwable) {
                        mainView?.onResponseFailure()
                        mainView?.hideProgress()
                        e.stackTrace
                    }
                })
        )
    }
}
